File: 6-kyu-kata/format-a-string-of-names-like-bart-lisa-and-maggie/python/format_a_string_of_names_like_bart_lisa_and_maggie_test.py
# 6 kyu - Format a string of names like 'Bart, Lisa & Maggie'.  [ ID: 53368a47e38700bd8300030d  (format-a-string-of-names-like-bart-lisa-and-maggie) ]
# URL: https://www.codewars.com/kata/53368a47e38700bd8300030d
# Category: REFERENCE  |  Tags: FUNDAMENTALS | STRINGS | DATA TYPES | FORMATTING | ALGORITHMS | LOGIC
# ******************************************************************************
def foo():

    from copy import deepcopy

    test.describe("Basic tests")
    test.assert_equals(
        namelist([{"name": "Bart"}, {"name": "Lisa"}, {"name": "Maggie"}, {"name": "Homer"}, {"name": "Marge"}]),
        "Bart, Lisa, Maggie, Homer & Marge",
        "Must work with many names",
    )
    test.assert_equals(
        namelist([{"name": "Bart"}, {"name": "Lisa"}, {"name": "Maggie"}]),
        "Bart, Lisa & Maggie",
        "Must work with many names",
    )
    test.assert_equals(namelist([{"name": "Bart"}, {"name": "Lisa"}]), "Bart & Lisa", "Must work with two names")
    test.assert_equals(namelist([{"name": "Bart"}]), "Bart", "Wrong output for a single name")
    test.assert_equals(namelist([]), "", "Must work with no names")

    test.describe("Random tests")
    from random import randint, shuffle

    def solist(names):
        return (lambda z: ", ".join(z[:-1]) + " & " + z[-1] if len(z) > 1 else "".join(z))(
            (lambda y: [x["name"] for x in y])(names)
        )

    base = [
        {"name": "Bart"},
        {"name": "Lisa"},
        {"name": "Maggie"},
        {"name": "Homer"},
        {"name": "Marge"},
        {"name": "Moe"},
        {"name": "Barney"},
        {"name": "Maude"},
        {"name": "Ned"},
        {"name": "Seymour"},
    ]
    for _ in range(40):
        shuffle(base)
        testmat = base[: randint(0, 9)]
        solution = solist(testmat)
        test.it("Testing for " + (solution if solution != "" else "None"))
        test.assert_equals(namelist(deepcopy(testmat)), solution, "It should work for random tests too")


foo()

File: kata-6-kyu/format-a-string-of-names-like-bart-lisa-and-maggie/javascript/formatAStringOfNamesLikeBartLisaAndMaggie.Test.js
//+ ====================================================================================================================
//+
//+ 6 kyu - Format a string of names like 'Bart, Lisa & Maggie'.  [ ID: 53368a47e38700bd8300030d ] (format-a-string-of-names-like-bart-lisa-and-maggie)
//+ URL: https://www.codewars.com/kata/53368a47e38700bd8300030d
//+ Category: REFERENCE  |  Tags: FUNDAMENTALS | STRINGS | DATA TYPES | FORMATTING | ALGORITHMS | LOGIC
//+
//+ ====================================================================================================================

const { assert } = require("chai")
const { list } = require("./formatAStringOfNamesLikeBartLisaAndMaggie")

Array.prototype.shuffle = function () {
  for (let i = this.length, j, c; i; ) {
    j = Math.floor(Math.random() * this.length)
    c = this[--i]
    this[i] = this[j]
    this[j] = c
  }
}

it("Basic tests", function () {
  assert.strictEqual(
    list([{ name: "Bart" }, { name: "Lisa" }, { name: "Maggie" }, { name: "Homer" }, { name: "Marge" }]),
    "Bart, Lisa, Maggie, Homer & Marge",
    "Must work with many names"
  )
  assert.strictEqual(
    list([{ name: "Bart" }, { name: "Lisa" }, { name: "Maggie" }]),
    "Bart, Lisa & Maggie",
    "Must work with many names"
  )
  assert.strictEqual(list([{ name: "Bart" }, { name: "Lisa" }]), "Bart & Lisa", "Must work with two names")
  assert.strictEqual(list([{ name: "Bart" }]), "Bart", "Wrong output for a single name")
  assert.strictEqual(list([]), "", "Must work with no names")
})

describe("Random tests", function () {
  function randint(a, b) {
    return Math.floor(Math.random() * (b - a + 1)) + a
  }
  function solist(names) {
    const res = []
    for (let item = 0; item < names.length; item++) {
      res.push(names[item].name)
    }
    return res.length > 1 ? res.slice(0, res.length - 1).join(", ") + " & " + res[res.length - 1] : res.join("")
  }
  let base = [
    { name: "Bart" },
    { name: "Lisa" },
    { name: "Maggie" },
    { name: "Homer" },
    { name: "Marge" },
    { name: "Moe" },
    { name: "Barney" },
    { name: "Maude" },
    { name: "Ned" },
    { name: "Seymour" }
  ]
  for (let i = 0; i < 40; i++) {
    base.shuffle()
    let testmat = base.slice(0, randint(0, 9))
    let solution = solist(testmat)
    it("Testing for " + (solution != "" ? solution : "None"), function () {
      assert.strictEqual(list(testmat), solution, "It should work for random tests too")
    })
  }
})
